base/abstractarray.jl

```diff
@@ -1964,6 +1983,358 @@ function typed_hvcat(::Type{T}, rows::Tuple{Vararg{Int}}, as...) where T
+    A = Array{T, N}(undef, dims...)
+    lengtha = length(A)  # Necessary to store result because throw blocks are being deoptimized right now, which leads to excessive allocations
+    lengthx = length(xs) # Cuts from 3 allocations to 1.
+    if lengtha != lengthx
+       throw(ArgumentError("argument count does not match specified shape (expected $lengtha, got $lengthx)"))
+    end
+    hvncat_fill!(A, row_first, xs)
+    return A
+end
+
+function hvncat_fill!(A::Array, row_first::Bool, xs::Tuple)
+    # putting these in separate functions leads to unnecessary allocations
+    if row_first
+        nr, nc = size(A, 1), size(A, 2)
+        nrc = nr * nc
+        @inbounds na = prod(size(A)[3:end])
+        k = 1
+        @inbounds for d ∈ 1:na
+            dd = nrc * (d - 1)
+            for i ∈ 1:nr
+                Ai = dd + i
+                for j ∈ 1:nc
+                    A[Ai] = xs[k]
+                    k += 1
+                    Ai += nr
+                end
+            end
+        end
+    else
+        @inbounds for k ∈ eachindex(xs)
+            A[k] = xs[k]
+        end
+    end
+end
+
+_typed_hvncat(T::Type, dim::Int, ::Bool, xs...) = _typed_hvncat(T, Val(dim), xs...) # catches from _hvncat type promoters
+_typed_hvncat(::Type{T}, ::Val) where T = Vector{T}()
+_typed_hvncat(T::Type, ::Val{N}, xs::Number...) where N = _typed_hvncat(T, (ntuple(x -> 1, N - 1)..., length(xs)), false, xs...)
+function _typed_hvncat(::Type{T}, ::Val{N}, as::AbstractArray...) where {T, N}
+    # optimization for arrays that can be concatenated by copying them linearly into the destination
+    # conditions: the elements must all have 1- or 0-length dimensions above N
+    @inbounds for a ∈ as
+        ndims(a) <= N || all(x -> size(a, x) == 1, (N + 1):ndims(a)) ||
+            return _typed_hvncat(T, (ntuple(x -> 1, N - 1)..., length(as)), false, as...)
+    end
+
+    nd = max(N, ndims(as[1]))
+
+    Ndim = 0
+    @inbounds for i ∈ 1:lastindex(as)
+        Ndim += cat_size(as[i], N)
+        for d ∈ 1:N - 1
+            cat_size(as[1], d) == cat_size(as[i], d) || throw(ArgumentError("mismatched size along axis $d in element $i"))
+        end
+    end
+
+    @inbounds A = Array{T, nd}(undef, ntuple(d -> cat_size(as[1], d), N - 1)..., Ndim, ntuple(x -> 1, nd - N)...)
+    k = 1
+    @inbounds for a ∈ as
+        for i ∈ eachindex(a)
+            A[k] = a[i]
+            k += 1
+        end
+    end
+    return A
+end
+
+# I don't understand why this function performs better when the first vararg is a number vs. an array
+# const y = fill(2)
+# _typed_hvncat(Int, Val(3), 2, y, y, y): 1 allocations and 56 ns
+# _typed_hvncat(Int, Val(3), y, y, y, 2): 2 allocations and 1.2 μs
+# both are still better than the (1, 1, 1, 2) argument variant
+# Apparently an issue with triggering specialization? #36307
+function _typed_hvncat(::Type{T}, ::Val{N}, as...) where {T, N}
```

Comment:
```suggestion
function _typed_hvncat(::Type{T}, ::Val{N}, as...) where {T, N}
```

This comment seems to be out of date?  On master on my machine I see:

<details>
<summary> set up</summary>

```Julia
$ julia-master
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.7.0-DEV.872 (2021-04-08)
 _/ |\__'_|_|_|\__'_|  |  Commit 027bd4a4d1 (4 days old master)
|__/                   |

julia> using Base: cat_size

julia> function _typed_hvncat(::Type{T}, ::Val{N}, as...) where {T, N}
            # optimization for scalars and 1-length arrays that can be concatenated by copying them linearly
            # into the destination
            nd = N
            Ndim = 0
            @inbounds for a ∈ as
                if a isa AbstractArray
                    cat_size(a, N) == length(a) ||
                        throw(ArgumentError("all dimensions of elements other than $N must be of length 1"))
                    nd = max(nd, ndims(a))
                end
                Ndim += cat_size(a, N)
            end

            @inbounds A = Array{T, nd}(undef, ntuple(x -> 1, N - 1)..., Ndim, ntuple(x -> 1, nd - N)...)

            k = 1
            @inbounds for a ∈ as
                if a isa AbstractArray
                    lena = length(a)
                    copyto!(A, k, a, 1, lena)
                    k += lena
                else
                    A[k] = a
                    k += 1
                end
            end
            return A
        end
_typed_hvncat (generic function with 1 method)

julia> using BenchmarkTools

julia> const y = fill(2)
0-dimensional Array{Int64, 0}:
2
```
</details>

```Julia
julia> @btime _typed_hvncat(Int, Val(3), 2, y, y, y);
  1.007 μs (2 allocations: 128 bytes)

julia> @btime _typed_hvncat(Int, Val(3), y, y, y, 2);
  1.007 μs (2 allocations: 128 bytes)

julia> @btime _typed_hvncat(Int, Val(3), 1, 1, 1, 2);
  37.851 ns (1 allocation: 112 bytes)
```

So: hurrah, we're consistently slow in the heterogeneous case?  In any case, better to track this in a separate issue than in a code comment.

Reply:
Yep, meant to remove, thanks!